File: cpp-wca-stats/src/loader.cpp
// Read a WCA export zip with libzip and fill WcaData.

#include "loader.hpp"

#include <algorithm>
#include <charconv>
#include <cstring>
#include <format>
#include <memory>
#include <ranges>
#include <stdexcept>
#include <string>
#include <string_view>
#include <vector>

#include <zip.h>

namespace wca {

namespace {

using std::string_view;

// Parse a signed integer; returns def on failure.  C++17+ std::from_chars
// avoids locale overhead that std::stoi has.
template <class T>
constexpr T parse_int(string_view s, T def = 0) noexcept {
    T v{};
    auto [p, ec] = std::from_chars(s.data(), s.data() + s.size(), v);
    return ec == std::errc{} ? v : def;
}

// Split a TSV line into fields.  The caller must keep the line alive.
struct TsvLine {
    std::vector<string_view> fields;
    void split(string_view line) {
        fields.clear();
        std::size_t start = 0;
        for (std::size_t i = 0; i < line.size(); ++i) {
            if (line[i] == '\t') {
                fields.emplace_back(line.data() + start, i - start);
                start = i + 1;
            }
        }
        fields.emplace_back(line.data() + start, line.size() - start);
    }
    string_view at(std::size_t i) const {
        return i < fields.size() ? fields[i] : string_view{};
    }
};

// Read the full contents of a named zip entry into a string.
std::string read_entry(zip_t* z, const char* name) {
    zip_stat_t st;
    zip_stat_init(&st);
    if (zip_stat(z, name, 0, &st) != 0) {
        throw std::runtime_error(std::format("missing zip entry: {}", name));
    }
    auto size = st.size;
    std::string buf;
    buf.resize(size);
    zip_file_t* f = zip_fopen(z, name, 0);
    if (!f) throw std::runtime_error(std::format("open entry: {}", name));
    std::size_t off = 0;
    while (off < size) {
        auto n = zip_fread(f, buf.data() + off, size - off);
        if (n <= 0) break;
        off += static_cast<std::size_t>(n);
    }
    zip_fclose(f);
    buf.resize(off);
    return buf;
}

// Iterate lines of a TSV (no embedded quoted newlines), skipping the header.
template <class Fn>
void for_each_data_line(string_view text, Fn fn) {
    std::size_t pos = 0;
    bool header_skipped = false;
    while (pos < text.size()) {
        auto nl = text.find('\n', pos);
        auto end = (nl == string_view::npos) ? text.size() : nl;
        auto line = text.substr(pos, end - pos);
        if (!line.empty() && line.back() == '\r') line.remove_suffix(1);
        if (!header_skipped) {
            header_skipped = true;
        } else if (!line.empty()) {
            fn(line);
        }
        pos = end + 1;
    }
}

// Parse the header row into a name->index map.
std::vector<string_view> header_fields(string_view text) {
    auto nl = text.find('\n');
    auto line = text.substr(0, nl == string_view::npos ? text.size() : nl);
    if (!line.empty() && line.back() == '\r') line.remove_suffix(1);
    TsvLine t; t.split(line);
    return t.fields;
}

}  // namespace

std::optional<std::uint16_t> WcaData::event_id(string_view name) const {
    if (auto it = event_idx.find(std::string{name}); it != event_idx.end()) return it->second;
    return std::nullopt;
}

std::optional<std::uint32_t> WcaData::person_key(string_view wca_id) const {
    if (auto it = person_idx_by_wca_id.find(std::string{wca_id});
        it != person_idx_by_wca_id.end()) return it->second;
    return std::nullopt;
}

std::vector<std::int32_t> WcaData::event_years(std::uint16_t eid) const {
    std::vector<std::uint8_t> seen(competitions.size(), 0);
    std::int32_t lo = INT32_MAX, hi = INT32_MIN;
    for (const auto& r : results) {
        if (r.event_id != eid) continue;
        if (!seen[r.comp_key]) {
            seen[r.comp_key] = 1;
            auto y = competitions[r.comp_key].year;
            lo = std::min(lo, y);
            hi = std::max(hi, y);
        }
    }
    std::vector<std::int32_t> out;
    if (lo <= hi) {
        out.reserve(static_cast<std::size_t>(hi - lo + 1));
        for (std::int32_t y = lo; y <= hi; ++y) out.push_back(y);
    }
    return out;
}

std::vector<std::int32_t> WcaData::person_event_years(std::uint32_t pk,
                                                      std::uint16_t eid) const {
    std::int32_t lo = INT32_MAX, hi = INT32_MIN;
    std::int32_t plo = INT32_MAX;
    for (const auto& r : results) {
        if (r.event_id != eid) continue;
        auto y = competitions[r.comp_key].year;
        lo = std::min(lo, y);
        hi = std::max(hi, y);
        if (r.person_key == pk) plo = std::min(plo, y);
    }
    std::vector<std::int32_t> out;
    if (plo != INT32_MAX && lo <= hi) {
        for (std::int32_t y = plo; y <= hi; ++y) out.push_back(y);
    }
    return out;
}

WcaData load_wca(const std::filesystem::path& zip_path) {
    auto fname = zip_path.filename().string();
    if (fname.find("WCA_export_v2_") == std::string::npos) {
        throw std::runtime_error("not a WCA v2 export zip");
    }

    int err = 0;
    std::unique_ptr<zip_t, decltype(&zip_close)> z{
        zip_open(zip_path.string().c_str(), ZIP_RDONLY, &err), &zip_close};
    if (!z) throw std::runtime_error(std::format("zip_open failed: {}", err));

    WcaData d;

    // ---- persons ----  (cols: name, gender, wca_id, sub_id, country_id)
    {
        auto body = read_entry(z.get(), "WCA_export_persons.tsv");
        TsvLine t;
        for_each_data_line(body, [&](string_view line) {
            t.split(line);
            auto wca_id = std::string{t.at(2)};
            auto sub_id = parse_int<std::uint16_t>(t.at(3), 1);
            if (sub_id == 1) {
                d.person_idx_by_wca_id.emplace(wca_id, static_cast<std::uint32_t>(d.persons.size()));
            }
            d.persons.push_back(Person{
                .wca_id     = std::move(wca_id),
                .sub_id     = sub_id,
                .name       = std::string{t.at(0)},
                .country_id = std::string{t.at(4)},
                .gender     = std::string{t.at(1)},
            });
        });
    }

    // ---- competitions ----
    {
        auto body = read_entry(z.get(), "WCA_export_competitions.tsv");
        auto hdr  = header_fields(body);
        auto col  = [&](string_view n) -> std::size_t {
            auto it = std::ranges::find(hdr, n);
            return static_cast<std::size_t>(it - hdr.begin());
        };
        const auto c_id   = col("id");
        const auto c_year = col("year");

        TsvLine t;
        for_each_data_line(body, [&](string_view line) {
            t.split(line);
            if (t.fields.size() <= c_year) return;
            auto id = std::string{t.at(c_id)};
            auto year = parse_int<std::int32_t>(t.at(c_year), 0);
            d.comp_idx_by_id.emplace(id, static_cast<std::uint32_t>(d.competitions.size()));
            d.competitions.push_back(Competition{.id = std::move(id), .year = year});
        });
    }

    // ---- results ---- (columns at fixed positions matching the Rust port)
    {
        auto body = read_entry(z.get(), "WCA_export_results.tsv");
        TsvLine t;
        for_each_data_line(body, [&](string_view line) {
            t.split(line);
            auto id      = parse_int<std::int64_t>(t.at(0), -1);
            if (id < 0) return;
            auto pos     = parse_int<std::int32_t>(t.at(1), 0);
            auto best    = parse_int<std::int32_t>(t.at(2), 0);
            auto average = parse_int<std::int32_t>(t.at(3), 0);
            auto comp_id = t.at(4);
            auto round   = t.at(5);
            auto ev      = t.at(6);
            auto person  = t.at(8);

            auto comp_it = d.comp_idx_by_id.find(std::string{comp_id});
            if (comp_it == d.comp_idx_by_id.end()) return;
            auto person_it = d.person_idx_by_wca_id.find(std::string{person});
            if (person_it == d.person_idx_by_wca_id.end()) return;

            std::uint16_t eid;
            if (auto eit = d.event_idx.find(std::string{ev}); eit != d.event_idx.end()) {
                eid = eit->second;
            } else {
                eid = static_cast<std::uint16_t>(d.events.size());
                d.events.emplace_back(ev);
                d.event_idx.emplace(std::string{ev}, eid);
            }

            d.results.push_back(Result333{
                .id            = id,
                .pos           = pos,
                .best          = best,
                .average       = average,
                .comp_key      = comp_it->second,
                .round_type_id = std::string{round},
                .person_key    = person_it->second,
                .event_id      = eid,
            });
        });
    }

    // ---- result_attempts ---- (keep only those referencing a kept result)
    {
        std::unordered_map<std::int64_t, std::uint8_t> wanted;
        wanted.reserve(d.results.size());
        for (const auto& r : d.results) wanted.emplace(r.id, 1);

        auto body = read_entry(z.get(), "WCA_export_result_attempts.tsv");
        TsvLine t;
        for_each_data_line(body, [&](string_view line) {
            t.split(line);
            auto value  = parse_int<std::int32_t>(t.at(0), INT32_MIN);
            if (value == INT32_MIN) return;
            auto ano    = parse_int<std::int32_t>(t.at(1), -1);
            if (ano < 0) return;
            auto rid    = parse_int<std::int64_t>(t.at(2), -1);
            if (rid < 0) return;
            if (!wanted.contains(rid)) return;
            d.attempts_by_result[rid].push_back(Attempt{
                .result_id      = rid,
                .attempt_number = static_cast<std::uint8_t>(ano),
                .value          = value,
            });
        });
        for (auto& [_, v] : d.attempts_by_result) {
            std::ranges::sort(v, {}, &Attempt::attempt_number);
        }
    }

    return d;
}

}  // namespace wca

File: rust-wca-stats/src/loader.rs
// Load WCA export zip: persons, competitions, results (filtered to a chosen
// event), and result_attempts (only for the kept result ids).

use ahash::{AHashMap, AHashSet};
use std::fs::File;
use std::io::{BufRead, BufReader, Read};
use std::path::Path;

pub struct Person {
    pub wca_id: String,
    pub sub_id: u16,
    pub name: String,
    pub country_id: String,
    pub gender: String,
}

pub struct Competition {
    pub id: String,
    pub year: i32,
}

pub struct Result333 {
    pub id: i64,
    pub pos: i32,
    pub best: i32,
    pub average: i32,
    pub comp_key: u32,
    pub round_type_id: String,
    pub person_key: u32, // points into persons (sub_id==1 entry)
    pub event_id: u16,
}

pub struct Attempt {
    pub result_id: i64,
    pub attempt_number: u8,
    pub value: i32,
}

pub struct WcaData {
    pub persons: Vec<Person>,              // all persons (all sub_ids)
    pub person_idx_by_wca_id: AHashMap<String, u32>, // wca_id -> index of sub_id==1 entry
    pub competitions: Vec<Competition>,
    pub comp_idx_by_id: AHashMap<String, u32>,
    pub events: Vec<String>,
    pub event_idx: AHashMap<String, u16>,
    pub results: Vec<Result333>,
    pub attempts_by_result: AHashMap<i64, Vec<Attempt>>, // sorted by attempt_number
}

impl WcaData {
    pub fn event_id(&self, name: &str) -> Option<u16> {
        self.event_idx.get(name).copied()
    }
    pub fn person_key(&self, wca_id: &str) -> Option<u32> {
        self.person_idx_by_wca_id.get(wca_id).copied()
    }

    pub fn event_years(&self, event_id: u16) -> Vec<i32> {
        let mut comp_ids: AHashSet<u32> = AHashSet::new();
        for r in &self.results {
            if r.event_id == event_id {
                comp_ids.insert(r.comp_key);
            }
        }
        let (mut lo, mut hi) = (i32::MAX, i32::MIN);
        for cid in &comp_ids {
            let y = self.competitions[*cid as usize].year;
            if y < lo { lo = y; }
            if y > hi { hi = y; }
        }
        (lo..=hi).collect()
    }

    pub fn person_event_years(&self, person_key: u32, event_id: u16) -> Vec<i32> {
        let mut all_comp: AHashSet<u32> = AHashSet::new();
        let mut person_comp: AHashSet<u32> = AHashSet::new();
        for r in &self.results {
            if r.event_id != event_id {
                continue;
            }
            all_comp.insert(r.comp_key);
            if r.person_key == person_key {
                person_comp.insert(r.comp_key);
            }
        }
        let mut lo = i32::MAX;
        for c in &person_comp {
            lo = lo.min(self.competitions[*c as usize].year);
        }
        let mut hi = i32::MIN;
        for c in &all_comp {
            hi = hi.max(self.competitions[*c as usize].year);
        }
        (lo..=hi).collect()
    }
}

pub fn load_wca(zip_path: &Path) -> anyhow::Result<WcaData> {
    let name = zip_path.file_name().unwrap().to_string_lossy().to_string();
    anyhow::ensure!(
        name.contains("WCA_export_v2_"),
        "not a WCA v2 export zip"
    );
    let f = File::open(zip_path)?;
    let mut zip = zip::ZipArchive::new(f)?;

    // persons
    let mut persons = Vec::new();
    let mut person_idx: AHashMap<String, u32> = AHashMap::new();
    {
        let file = zip.by_name("WCA_export_persons.tsv")?;
        let mut rdr = make_reader(file);
        let mut rec = csv::ByteRecord::new();
        // header row consumed by csv reader
        while rdr.read_byte_record(&mut rec)? {
            let name = field(&rec, 0);
            let gender = field(&rec, 1);
            let wca_id = field(&rec, 2);
            let sub_id: u16 = field(&rec, 3).parse().unwrap_or(1);
            let country = field(&rec, 4);
            if sub_id == 1 {
                person_idx.insert(wca_id.to_string(), persons.len() as u32);
            }
            persons.push(Person {
                name: name.to_string(),
                gender: gender.to_string(),
                wca_id: wca_id.to_string(),
                sub_id,
                country_id: country.to_string(),
            });
        }
    }

    // competitions (we only need id + year). Competitions tsv has many fields
    // and some free-text ones, but row boundaries are line breaks and fields
    // are plain tab-separated; columns we want are fixed-position by header.
    let mut competitions = Vec::new();
    let mut comp_idx: AHashMap<String, u32> = AHashMap::new();
    {
        let file = zip.by_name("WCA_export_competitions.tsv")?;
        let mut rdr = BufReader::with_capacity(1 << 20, file);
        let mut header = String::new();
        rdr.read_line(&mut header)?;
        let header_fields: Vec<&str> = header.trim_end_matches('\n').split('\t').collect();
        let id_col = header_fields.iter().position(|x| *x == "id").unwrap();
        let year_col = header_fields.iter().position(|x| *x == "year").unwrap();
        let mut line = String::new();
        loop {
            line.clear();
            let n = rdr.read_line(&mut line)?;
            if n == 0 { break; }
            let line_t = line.trim_end_matches('\n');
            let fs: Vec<&str> = line_t.split('\t').collect();
            if fs.len() <= year_col { continue; }
            let id = fs[id_col].to_string();
            let year: i32 = fs[year_col].parse().unwrap_or(0);
            comp_idx.insert(id.clone(), competitions.len() as u32);
            competitions.push(Competition { id, year });
        }
    }

    // events: build from results as they are loaded (unique event ids).
    let mut events: Vec<String> = Vec::new();
    let mut event_idx: AHashMap<String, u16> = AHashMap::new();

    // results
    let mut results: Vec<Result333> = Vec::new();
    {
        let file = zip.by_name("WCA_export_results.tsv")?;
        let mut rdr = make_reader(file);
        let mut rec = csv::ByteRecord::new();
        // id, pos, best, average, competition_id, round_type_id, event_id,
        // person_name, person_id, format_id, regional_single_record,
        // regional_average_record, person_country_id
        while rdr.read_byte_record(&mut rec)? {
            let id: i64 = field(&rec, 0).parse()?;
            let pos: i32 = field(&rec, 1).parse().unwrap_or(0);
            let best: i32 = field(&rec, 2).parse().unwrap_or(0);
            let average: i32 = field(&rec, 3).parse().unwrap_or(0);
            let comp_id = field(&rec, 4);
            let round_type = field(&rec, 5).to_string();
            let ev = field(&rec, 6);
            let person_id = field(&rec, 8);
            let comp_key = match comp_idx.get(comp_id) {
                Some(k) => *k,
                None => continue,
            };
            let person_key = match person_idx.get(person_id) {
                Some(k) => *k,
                None => continue,
            };
            let event_id = match event_idx.get(ev) {
                Some(&k) => k,
                None => {
                    let k = events.len() as u16;
                    events.push(ev.to_string());
                    event_idx.insert(ev.to_string(), k);
                    k
                }
            };
            results.push(Result333 {
                id,
                pos,
                best,
                average,
                comp_key,
                round_type_id: round_type,
                person_key,
                event_id,
            });
        }
    }

    // result_attempts: only keep attempts whose result_id is in the set of
    // result ids for events we care about. For generality (we support any
    // event via calc), keep all attempts whose result_id is in our results.
    let wanted: AHashSet<i64> = results.iter().map(|r| r.id).collect();
    let mut attempts_by_result: AHashMap<i64, Vec<Attempt>> = AHashMap::new();
    {
        let file = zip.by_name("WCA_export_result_attempts.tsv")?;
        let mut rdr = make_reader(file);
        let mut rec = csv::ByteRecord::new();
        while rdr.read_byte_record(&mut rec)? {
            let value: i32 = match field(&rec, 0).parse() { Ok(v) => v, Err(_) => continue };
            let attempt_number: u8 = match field(&rec, 1).parse() { Ok(v) => v, Err(_) => continue };
            let result_id: i64 = match field(&rec, 2).parse() { Ok(v) => v, Err(_) => continue };
            if !wanted.contains(&result_id) { continue; }
            attempts_by_result
                .entry(result_id)
                .or_default()
                .push(Attempt { result_id, attempt_number, value });
        }
    }
    for v in attempts_by_result.values_mut() {
        v.sort_by_key(|a| a.attempt_number);
    }

    Ok(WcaData {
        persons,
        person_idx_by_wca_id: person_idx,
        competitions,
        comp_idx_by_id: comp_idx,
        events,
        event_idx,
        results,
        attempts_by_result,
    })
}

fn make_reader<R: Read>(r: R) -> csv::Reader<BufReader<R>> {
    csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .quoting(false)
        .has_headers(true)
        .flexible(true)
        .from_reader(BufReader::with_capacity(1 << 20, r))
}

fn field<'a>(rec: &'a csv::ByteRecord, i: usize) -> &'a str {
    std::str::from_utf8(rec.get(i).unwrap_or(b"")).unwrap_or("")
}
